### PR TITLE
Add skillsaw port: convert Claude/Codex plugins to Agent Plugins v1

### PR DESCRIPTION
**Pull request by not-stbenjam**
Adds `skillsaw port` — one command to make every plugin in a repository installable by any Agent Plugins v1 client — plus the opt-in `agent-plugin-required` rule that keeps it that way. Targeted at 0.18.0: it makes adopting the vendor-neutral standard a four-second operation.

```
$ skillsaw port --to agent-plugin .
skillsaw 0.18.0
Porting to Agent Plugins v1: /repos/my-marketplace

  ✓ [plugins/release-notes] claude → plugin.json, mcp.json
  ✓ [plugins/issue-triage] codex → plugin.json
      note: commands/ stay as client-specific content — Agent Plugins v1 defines only skills and MCP servers

Summary:
  Ported:     2 plugins (2 plugin.json, 1 mcp.json)
  Validation: ✓ passed

✓ Converted 2 plugins to Agent Plugins v1
```

## Design

- **Additive by construction.** Agent Plugins v1 coexists with client-specific content (spec §8), so a port only ever writes root `plugin.json` and `mcp.json`. Claude/Codex files are never modified; commands, agents, and hooks stay behind as client-specific content with a note.
- **Recursive.** Reuses lint discovery (`distinct_plugin_dirs`), so a marketplace ports all its plugins in one run and a single-plugin repo ports its root. Interactive runs get a `converting [N/total]` progress line.
- **Spec-faithful translation** (implemented against the canonical spec text, rev in `SCHEMA-SOURCE.md`): manifest metadata with §5.5 name normalization; MCP `http` → `streamable-http`, `${CLAUDE_PLUGIN_ROOT}` → `${PLUGIN_ROOT}` (or `./`-relative `command` per §7.2.1); anything the portable format cannot express (`ws` transport, shell-style commands, reserved env names) is skipped with a note, never mistranslated.
- **Self-verifying.** Every port ends with `agent-plugin-json-valid` / `agent-plugin-mcp-valid` run over the output; the command fails if its own output doesn't validate. A foreign root `plugin.json` is refused, never overwritten. Reruns are no-ops.
- One module (`src/skillsaw/port.py`, ~250 lines), plan-then-apply, no LLM, no network.

## The rule

`agent-plugin-required` (opt-in, warning) makes "all plugins must be available in the vendor-neutral format" enforceable in CI:

- flags plugins missing the portable manifest — **SAFE-fixable**: `skillsaw fix` runs the same planner (deterministic, additive-only, and the rule is opt-in so intent is explicit)
- flags shared-metadata drift between the manifests (reported, never auto-edited)
- flags a Claude `.mcp.json` with no portable `mcp.json` counterpart (fixable)

This is a builtin generalization of the custom rule in openshift-eng/ai-helpers#685, which can retire it.

## Validation

- **Ground truth**: ported `openshift-eng/ai-helpers` (40 plugins) and diffed against the hand-written manifests in openshift-eng/ai-helpers#685 — **37/40 `plugin.json` byte-identical** (3 differ only by a version bump made on that branch) and **3/3 `mcp.json` byte-identical**, transport translation included. Machine-generated comparison branch: https://github.com/stbenjam/ai-helpers/tree/skillsaw-port-demo
- **`openai/plugins` (180 Codex plugins)**: all 180 ported with 8 MCP translations, validation passed, and the repo's existing `.agents/plugins/marketplace.json` correctly left untouched. Spot-checks: manifests field-identical to their `.codex-plugin` sources, MCP configs translated verbatim with `type: stdio` added.
- `anthropics/claude-plugins-official`: 21 plugins ported, validation passed, rerun clean.
- 21 new tests (`tests/test_port.py`) with static fixtures: recursive marketplace port, dry-run, conflict refusal, idempotency, MCP edge cases, rule fire/fix/drift/mcp-counterpart; plus a coverage fixture so `agent-plugin-required` fires in the rule-coverage guard.
- Docs: `docs/porting.md` guide (in nav), long-form rule doc, generated pages, README row. `make test` (3890), `make update`, `make lint` all clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added `skillsaw port` to convert Claude Code and Codex plugins into vendor-neutral Agent Plugins v1 packages.
  * Supports dry runs, validation, progress controls, overwrite protection, and recursive conversion.
  * Added optional `agent-plugin-required` checks with autofix support for portable packaging, synchronized metadata, and MCP configuration.

* **Documentation**
  * Added CLI guidance and a dedicated guide for porting plugins to Agent Plugins.
  * Documented the new rule, supported conversions, skipped configurations, and validation behavior.

* **Bug Fixes**
  * Improved reporting for findings without file locations.

### CODE REVIEW

File: docs/porting.md [hunk 1]
```diff
@@ -52,6 +52,19 @@ output; the command fails if its own output doesn't validate. Use
 over an already-ported tree is a no-op, and a root `plugin.json` that
 belongs to something else is never overwritten.
 
+## Marketplace catalogs
+
+Agent Plugins v1 defines a package, not a marketplace — clients that
```
Comment (stbenjam) [inline docs/porting.md:57]: We should note here, the reason we are doing this is it seems likely Codex writing its marketplace into `.agents` means it was some attempt at a vendor neutral format-ish type thing, and it is a reasonable thing to do until there's some kind of standard

File: src/skillsaw/cli/_parser.py [hunk 2]
```diff
@@ -322,6 +322,46 @@ def _build_parser():
         "ocean-blue, sunset-orange, royal-purple, crimson-red.",
     )
 
+    # --- port ---
+    port_parser = subparsers.add_parser(
+        "port",
+        help="Port Claude Code and Codex plugins to Agent Plugins v1 packages",
+    )
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/cli/_parser.py:329]: The README's workflows table has a row for this ("Convert plugins to the vendor-neutral format" → the Porting guide), and `port` appears in the generated CLI reference. Happy to expand if you want more than the table row.

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    "keywords",
+)
+
+
+class AgentPluginRequiredRule(Rule):
+    """Require every Claude/Codex plugin to also be an Agent Plugins package."""
+
+    default_enabled = False
+    since = "0.18.0"
+    autofix_confidence = AutofixConfidence.SAFE
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py:38]: Declining: the rule is opt-in (`default_enabled = False`), and `repo_types` only gates `enabled: auto` activation — an explicit `enabled: true` must always run it. Declaring repo types here would add a gate with no default effect.

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            if v.file_path is None:
+                continue
+            plugin_dir = v.file_path.parent
+            if "not available as an Agent Plugins" in v.message:
+                plan = plan_port(plugin_dir)
+                for filename, content in plan.writes.items():
+                    results.append(
+                        AutofixResult(
+                            rule_id=self.rule_id,
+                            file_path=plugin_dir / filename,
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py:143]: No guard needed by design: the linter stands autofix down centrally for everything under `.codex/plugins/` (every rule's fix silently no-ops there), so per-rule guards are deliberately not written. This is pinned in the repo's development rules.

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        violations: List[RuleViolation] = []
+        for plugin_dir in context.distinct_plugin_dirs():
+            source, _sources, _notes = read_source_manifest(plugin_dir)
+            if source is None:
+                continue
+
+            if not declares_agent_plugin(plugin_dir):
+                violations.append(
+                    self.violation(
+                        "Plugin is not available as an Agent Plugins v1 package — "
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py]: Fixed: violations now compute fixability from the planner — missing-manifest violations advertise a fix only when `plan_port()` produces writes, and the MCP counterpart only when at least one server survives translation.

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        for field in _SHARED_FIELDS:
+            if field not in source or field not in portable:
+                continue
+            left, right = source[field], portable[field]
+            if field == "name" and isinstance(left, str):
+                left = normalize_name(left) or left
+            if field == "author" and isinstance(left, str):
+                left = {"name": left}
+            if left != right:
+                drifted.append(field)
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py]: Fixed: `_drifted_fields` now normalizes the source author exactly as the port renders it (string → object, non-portable keys dropped), so a clean port never reads as drift.

File: src/skillsaw/port.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        if not isinstance(command, str) or not command:
+            notes.append(f"MCP server '{name}' has no usable 'command' — skipped")
+            return None
+        # Spec §7.2.1: command is one executable token, bare or ./-relative,
+        # with no placeholder expansion. The Claude root variable at the
+        # front of a path is exactly a plugin-relative reference.
+        if command.startswith(_CLAUDE_ROOT_VAR + "/"):
+            command = "./" + command[len(_CLAUDE_ROOT_VAR) + 1 :]
+        if _CLAUDE_ROOT_VAR in command or " " in command:
+            notes.append(
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/port.py:274]: Fixed: the command check now rejects any whitespace (`any(ch.isspace() ...)`), matching the validator.

File: src/skillsaw/port.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    _note_client_specific_content(root, plan.notes)
+    return plan
+
+
+def apply_plan(plan: PortPlan) -> List[Path]:
+    """Write a plan's files. Returns the paths written."""
+    written = []
+    for filename, content in plan.writes.items():
+        path = plan.root / filename
+        path.write_text(content, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/port.py:130]: Fixed: symlinked `plugin.json`/`mcp.json` targets (including dangling ones) are refused at plan time, and `apply_plan` re-checks `is_symlink()` at write time.

File: src/skillsaw/cli/_port.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                file_counts[filename] = file_counts.get(filename, 0) + 1
+
+            if args.dry_run:
+                continue
+
+            apply_plan(plan)
+            # A port isn't done until the target format's own rules accept
+            # it. Config advisories (e.g. deprecation notices) belong to
+            # lint runs; only the format's own findings gate the port.
+            context = RepositoryContext(target, exclude_patterns=config.exclude_patterns)
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/cli/_port.py:135]: Fixed — and it was worse than stale results: the cached miss meant discovery never saw the new manifest, so in-process verification passed vacuously. The CLI now drops the read caches after writing before building the verification context. (The fixture test validated in a fresh process, which is why it stayed honest.)

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        return Severity.WARNING
+
+    def check(self, context: RepositoryContext) -> List[RuleViolation]:
+        violations: List[RuleViolation] = []
+        for plugin_dir in context.distinct_plugin_dirs():
+            source, _sources, _notes = read_source_manifest(plugin_dir)
+            if source is None:
+                continue
+
+            if not declares_agent_plugin(plugin_dir):
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py]: Fixed: the rule now answers ownership from discovery's verdicts — membership in `context.agent_plugins` — instead of probing the filesystem, per the ecosystem-provenance rule.

File: src/skillsaw/cli/_port.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    if verify_failures:
+        print(f"  Validation: {c['red']}✗ {len(verify_failures)} violation(s){c['reset']}")
+        print(
+            f"\n{c['red']}✗{c['reset']} Ported {len(ported)} {plugin_word}; fix the violations above."
+        )
+        sys.exit(1)
+
+    print(f"  Validation: {c['green']}✓ passed{c['reset']}")
+    print(f"\n{c['green']}✓{c['reset']} Converted {len(ported)} {plugin_word} to Agent Plugins v1")
+    sys.exit(0)
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/cli/_port.py:238]: Fixed: refused plugins now fail the run even when other plugins ported (`✗ Converted N plugins, but M plugin(s) were refused`), with a regression test.

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        drifted = []
+        for field in _SHARED_FIELDS:
+            if field not in source or field not in portable:
+                continue
+            left, right = source[field], portable[field]
+            if field == "name" and isinstance(left, str):
+                left = normalize_name(left) or left
+            if field == "author" and isinstance(left, str):
+                left = {"name": left}
+            if left != right:
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py]: Fixed along with the Qodo/CodeRabbit reports of the same: source author objects are filtered to name/email/url before comparison.

File: src/skillsaw/port.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+        plan.skipped = "no readable .claude-plugin/plugin.json or .codex-plugin/plugin.json"
+        return plan
+
+    existing = read_json(root / "plugin.json")[0]
+    if (root / "plugin.json").exists():
+        if isinstance(existing, dict) and is_agent_plugin_schema(existing.get("$schema"), "plugin"):
+            plan.skipped = "already an Agent Plugins package (plugin.json declares the schema)"
+        else:
+            plan.skipped = "a plugin.json not declaring the Agent Plugins schema already exists — refusing to overwrite"
+        return plan
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/port.py]: Fixed: a package that already declares Agent Plugins but lacks the portable `mcp.json` now gets the MCP translation instead of being skipped, with a regression test.

File: src/skillsaw/port.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+    return merged, sources, notes
+
+
+def read_source_mcp(root: Path) -> Optional[Dict[str, Any]]:
+    """Return the source ``mcpServers`` mapping, if the plugin has one."""
+    data, error = read_json(root / ".mcp.json")
+    if error or not isinstance(data, dict):
+        return None
+    servers = data.get("mcpServers")
+    return servers if isinstance(servers, dict) else None
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/port.py]: Deferred deliberately: the port reads the wrapped root `.mcp.json` both source formats share and stays conservative about other shapes — a flat server map or a Codex inline declaration can be added later without changing the plan model. Tracked as follow-up scope rather than mistranslating now.

File: src/skillsaw/port.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+    merged: Dict[str, Any] = {}
+    sources: List[str] = []
+    for label, rel in (
+        ("claude", ".claude-plugin/plugin.json"),
+        ("codex", ".codex-plugin/plugin.json"),
+    ):
+        path = root / rel
+        if not path.is_file():
+            continue
+        data, error = read_json(path)
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/port.py]: Fixed: source manifests and `.mcp.json` are only read when `contained_resolve()` keeps them inside the plugin root.

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+                            fixable=False,
+                        )
+                    )
+
+            if (plugin_dir / ".mcp.json").is_file() and not (plugin_dir / "mcp.json").is_file():
+                violations.append(
+                    self.violation(
+                        "Claude MCP configuration (.mcp.json) has no portable "
+                        "mcp.json counterpart",
+                        file_path=plugin_dir / "mcp.json",
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py:95]: Fixed: the MCP-counterpart violation only advertises a fix when `render_mcp` can actually produce a file.

File: tests/test_integration.py [hunk 17]
```diff
@@ -2751,7 +2752,9 @@ def _discover_safe_autofix_rule_ids() -> Set[str]:
     safe_ids: Set[str] = set()
     for rule_class in BUILTIN_RULES:
         instance = rule_class()
-        if instance.deprecated is not None:
+        # The fixture exercises a default lint run: deprecated rules no
+        # longer run in one, and opt-in rules never did.
+        if instance.deprecated is not None or instance.default_enabled is False:
             continue
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply tests/test_integration.py:2758]: Declining the revert: that fixture exercises a default lint run, and opt-in rules cannot fire in one regardless of confidence. The create-file autofix is covered directly in tests/test_port.py — fix creation, satisfied-after-fix re-check, and idempotent rerun.

File: scripts/generate-docs.py [hunk 18]
```diff
@@ -37,7 +37,7 @@
     ),
     (
         "Agent Plugins",
-        ["agent-plugin-json-valid", "agent-plugin-mcp-valid"],
+        ["agent-plugin-json-valid", "agent-plugin-mcp-valid", "agent-plugin-required"],
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply scripts/generate-docs.py:40]: Fixed: the five missing rules (coderabbit-schema-valid plus four content rules — pre-existing drift, invisible because the README no longer carries the generated markers so the script exits before its check) are back in the groups, aligned with the site generator.

File: src/skillsaw/cli/_port.py [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+                context,
+                config,
+                rule_ids=verify_rules,
+                no_plugins=True,
+                no_custom_rules=True,
+            )
+            for violation in linter.run():
+                if violation.rule_id in verify_rules:
+                    verify_failures.append(violation)
+                    print(f"  {violation}")
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/cli/_port.py:146]: Fixed together with the Codex report of the same: caches invalidated after writes, before the verification context is built.

File: src/skillsaw/port.py [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+
+PLUGIN_SCHEMA = "https://agent-plugins.org/schemas/1.0.0/plugin.schema.json"
+MCP_SCHEMA = "https://agent-plugins.org/schemas/1.0.0/mcp.schema.json"
+
+# Manifest metadata fields shared verbatim between the source formats and
+# the Agent Plugins closed schema (spec §5.4).
+_COPIED_STRING_FIELDS = ("version", "description", "homepage", "repository", "license")
+
+# Spec §5.5: 1-64 chars of [a-z0-9.-], alphanumeric ends, no -- or ..
+_NAME_RE = re.compile(r"^(?!.*(?:--|\.\.))[a-z0-9](?:[a-z0-9.-]*[a-z0-9])?$")
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/port.py]: Fixed: the lookahead is gone — the pattern matches the character shape greedily and the no-repetition rule (`--`/`..`) is checked in code, per the autofix-invariants rule.

File: src/skillsaw/rules/builtin/agent_plugins/required.py [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+            if field not in source or field not in portable:
+                continue
+            left, right = source[field], portable[field]
+            if field == "name" and isinstance(left, str):
+                left = normalize_name(left) or left
+            if field == "author" and isinstance(left, str):
+                left = {"name": left}
+            if left != right:
+                drifted.append(field)
+        return drifted
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/agent_plugins/required.py:127]: Fixed: source author normalized (string→object, non-portable keys dropped) before comparison, so a clean port reports no drift.

File: src/skillsaw/rules/docs/agent-plugin-required.md [hunk 22]
```diff
@@ -0,0 +1,54 @@
+## Why
+
+Agent Plugins v1 is the vendor-neutral plugin format: a root
+`plugin.json`, skills at `skills/*/SKILL.md`, and an optional portable
+`mcp.json`. It coexists with Claude Code and Codex formats in the same
+directory, so publishing it costs two generated files per plugin — and
+buys installation by any conforming client.
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/rules/docs/agent-plugin-required.md:7]: Fixed: the doc now says publishing costs at most two generated files, with mcp.json only when the plugin has MCP configuration; regenerated.

File: docs/porting.md (hunk 1 above)
Comment (not-stbenjam) [reply docs/porting.md:57]: Added to docs/porting.md: Codex writing its catalog into the client-neutral .agents/ directory reads as a step toward a vendor-neutral catalog format, making it the reasonable one to emit until a real standard exists.
